var/spack/repos/builtin/packages/perl-datetime-format-pg/package.py

```diff
@@ -0,0 +1,32 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PerlDatetimeFormatPg(PerlPackage):
+    """Parse and format PostgreSQL dates and times"""
+
+    homepage = "https://metacpan.org/pod/DateTime::Format::Pg"
+    url = "https://cpan.metacpan.org/authors/id/D/DM/DMAKI/DateTime-Format-Pg-0.16014.tar.gz"
+
+    maintainers("EbiArnie")
+
+    license("Artistic-1.0-Perl OR GPL-1.0-or-later")
+
+    version("0.16014", sha256="38bb9666524dc384c3366f6342cb9656c50bac0f9716a3d44f1cf552ccbe0eb9")
+
+    depends_on("perl-datetime@0.10:", type=("build", "run", "test"))
+    depends_on("perl-datetime-format-builder@0.72:", type=("build", "run", "test"))
+    depends_on("perl-datetime-timezone@0.05:", type=("build", "run", "test"))
+    depends_on("perl-module-build-tiny@0.035:", type=("build"))
+
+    def test_use(self):
+        """Test 'use module'"""
+        options = ["-we", 'use strict; use DateTime::Format::Pg; print("OK\n")']
+
+        perl = self.spec["perl"].command
+        out = perl(*options, output=str.split, error=str.split)
+        assert "OK" in out
```

Comment:
Variations of this pattern appear in a lot of PR you submitted, that add new Perl packages. Is it worth pushing this method to `PerlPackage` and let the specific package dictate only the variable parts[^1] ? Would it be ok if this implementation is shared among Perl packages (asking since I have little to no knowledge of Perl packaging format).

[^1]: as far as I can tell only the fully qualified name changes e.g. `use DateTime::Format::Pg`

Reply:
Yes, this should absolutely be in PerlPackage. Spack has a similar thing for Python.
And indeed the only difference is the Perl package name here.
And yes, this can be shared among all Perl packages.

The main reason I didn't put it there so far is that I didn't take the time to figure out where exactly this should go, PerlPackage or PerlBuilder and what else is needed.
Plus I would like this to be automatic, but could not think of a nice / easy way to find out the Perl package name from within the Spack PerlPackage superclass. We could still provide at least a nice wrapper method or annotation or something, as you suggested.

And this would then also need docs and again, I didn't take the time so far to look at that part of Spack.
And I don't know if Spack runs these test in CI, but if you do, there are probably a few broken Perl packages too.

I plan to do this eventually and then remove these from all my packages.

Reply:
> Plus I would like this to be automatic, but could not think of a nice / easy way to find out the Perl package name from within the Spack PerlPackage superclass.

If it was completely automatic, that would be :100: but even without that we can push the method up and ask each package to just fill an attribute with the Perl package name.